marlowe-playground-client/src/Marlowe/Holes.purs

```diff
@@ -0,0 +1,567 @@
+module Marlowe.Holes where
+
+import Prelude
+import Data.Array (foldMap, foldl, mapWithIndex, (:))
+import Data.BigInteger (BigInteger)
+import Data.Foldable (intercalate)
+import Data.FoldableWithIndex (foldlWithIndex)
+import Data.Generic.Rep (class Generic)
+import Data.Generic.Rep.Eq (genericEq)
+import Data.Generic.Rep.Show (genericShow)
+import Data.Map (Map)
+import Data.Map as Map
+import Data.Maybe (Maybe(..))
+import Data.String (length)
+import Data.Traversable (traverse)
+import Data.Tuple (Tuple(..))
+import Data.Tuple.Nested ((/\))
+import Marlowe.Pretty (class Pretty, genericPretty, prettyFragment)
+import Marlowe.Semantics (ChosenNum, Money, Party, PubKey, Slot, Timeout)
+import Marlowe.Semantics as S
+import Text.Parsing.Parser.Basic (replaceInPosition)
+import Text.Parsing.Parser.Pos (Position(..))
+import Text.PrettyPrint.Leijen (appendWithSoftbreak)
+import Text.PrettyPrint.Leijen as Leijen
+import Type.Proxy (Proxy)
+
+data MarloweType
+  = StringType
+  | BigIntegerType
+  | SlotType
+  | AccountIdType
+  | ChoiceIdType
+  | ValueIdType
+  | ActionType
+  | PayeeType
+  | CaseType
+  | ValueType
+  | InputType
+  | ObservationType
+  | ContractType
+  | BoundType
+
+derive instance eqMarloweType :: Eq MarloweType
+
+data Argument
+  = ArrayArg String
+  | DataArg String
+  | NewtypeArg
+
+getMarloweConstructors :: MarloweType -> Map String (Array Argument)
+getMarloweConstructors StringType = Map.singleton "String" [ NewtypeArg ]
+
+getMarloweConstructors BigIntegerType = Map.singleton "Integer" [ NewtypeArg ]
+
+getMarloweConstructors SlotType = Map.singleton "Integer" [ NewtypeArg ]
+
+getMarloweConstructors AccountIdType = Map.singleton "AccountId" [ DataArg "accNumber", DataArg "accHolder" ]
+
+getMarloweConstructors ChoiceIdType = Map.singleton "ChoiceId" [ DataArg "choiceNumber", DataArg "choiceOwner" ]
+
+getMarloweConstructors ValueIdType = Map.singleton "ValueId" [ NewtypeArg ]
+
+getMarloweConstructors ActionType =
+  Map.fromFoldable
+    [ (Tuple "Deposit" [ DataArg "accountId", DataArg "party", DataArg "value" ])
+    , (Tuple "Choice" [ DataArg "choiceId", ArrayArg "bounds" ])
+    , (Tuple "Notify" [ DataArg "observation" ])
+    ]
+
+getMarloweConstructors PayeeType =
+  Map.fromFoldable
+    [ (Tuple "Account" [ DataArg "accountId" ])
+    , (Tuple "Party" [ DataArg "party" ])
+    ]
+
+getMarloweConstructors CaseType = Map.singleton "Case" [ DataArg "action", DataArg "contract" ]
+
+getMarloweConstructors ValueType =
+  Map.fromFoldable
+    [ (Tuple "AvailableMoney" [ DataArg "accountId" ])
+    , (Tuple "Constant" [ DataArg "amount" ])
+    , (Tuple "NegValue" [ DataArg "value" ])
+    , (Tuple "AddValue" [ DataArg "value", DataArg "value" ])
+    , (Tuple "SubValue" [ DataArg "value", DataArg "value" ])
+    , (Tuple "ChoiceValue" [ DataArg "choiceId", DataArg "value" ])
+    , (Tuple "SlotIntervalStart" [])
+    , (Tuple "SlotIntervalEnd" [])
+    , (Tuple "UseValue" [ DataArg "valueId" ])
+    ]
```

Comment:
I have to wonder if this stuff could be generated with Generics. But that's a thought for another PR...

Reply:
I've already thought about this and as you say, I want to save it for another PR.